src/codeu/chat/common/Secret.java

```diff
@@ -14,23 +14,131 @@
+    TO_HEX.put(0x3, '3');
+    TO_HEX.put(0x4, '4');
+    TO_HEX.put(0x5, '5');
+    TO_HEX.put(0x6, '6');
+    TO_HEX.put(0x7, '7');
+    TO_HEX.put(0x8, '8');
+    TO_HEX.put(0x9, '9');
+    TO_HEX.put(0xA, 'A');
+    TO_HEX.put(0xB, 'B');
+    TO_HEX.put(0xC, 'C');
+    TO_HEX.put(0xD, 'D');
+    TO_HEX.put(0xE, 'E');
+    TO_HEX.put(0xF, 'F');
+
+    FROM_HEX.put('0', 0x0);
+    FROM_HEX.put('1', 0x1);
+    FROM_HEX.put('2', 0x2);
+    FROM_HEX.put('3', 0x3);
+    FROM_HEX.put('4', 0x4);
+    FROM_HEX.put('5', 0x5);
+    FROM_HEX.put('6', 0x6);
+    FROM_HEX.put('7', 0x7);
+    FROM_HEX.put('8', 0x8);
+    FROM_HEX.put('9', 0x9);
+    FROM_HEX.put('A', 0xA);
+    FROM_HEX.put('B', 0xB);
+    FROM_HEX.put('C', 0xC);
+    FROM_HEX.put('D', 0xD);
+    FROM_HEX.put('E', 0xE);
+    FROM_HEX.put('F', 0xF);
+    FROM_HEX.put('a', 0xA);
+    FROM_HEX.put('b', 0xB);
+    FROM_HEX.put('c', 0xC);
+    FROM_HEX.put('d', 0xD);
+    FROM_HEX.put('e', 0xE);
+    FROM_HEX.put('f', 0xF);
+  }
+
+  public static final Serializer<Secret> SERIALIZER = new Serializer<Secret>() {
+    @Override
+    public void write(OutputStream out, Secret value) throws IOException {
+      Serializers.BYTES.write(out, value.bytes);
+    }
+
+    @Override
+    public Secret read(InputStream in) throws IOException {
+      return new Secret(Serializers.BYTES.read(in));
+    }
+  };
+
+  private final byte[] bytes;
+
+  public Secret(byte... bytes) {
+    this.bytes = Arrays.copyOf(bytes, bytes.length);
+  }
+
+  @Override
+  public String toString() {
+    final StringBuilder string = new StringBuilder();
+
+    for (final byte b : bytes) {
+      string.append(TO_HEX.get((0xF0 & b) >>> 8));
+      string.append(TO_HEX.get((0x0F & b)));
+    }
+
+    return string.toString();
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    return other instanceof Secret ?
+        Arrays.equals(bytes, ((Secret)other).bytes) :
+        false;
+  }
+
+  @Override
+  public int hashCode() {
+    return Arrays.hashCode(bytes);
+  }
+
   // PARSE
   //
   // Take in the string representation of the secret and convert it into a byte
   // array. The string form of a secret should be a hex string that will be
   // converted into a byte array.
   //
   // For example: "ABABAB" becomes { 0xAB, 0xAB, 0xAB }
-  public static byte[] parse(String stringSecret) {
+  public static Secret parse(String stringSecret) throws IOException {
+
+    // Make sure the string only has valid hex characters.
+    for (int i = 0; i < stringSecret.length(); i++) {
+      if (FROM_HEX.get(stringSecret.charAt(i)) == null) {
+        throw new IOException(String.format(
+            "Unsupported character '%c'",
+            stringSecret.charAt(i)));
+      }
+    }
 
-    final byte[] expanded = new byte[stringSecret.length() + stringSecret.length() % 2];
+    final int[] expanded = new int[stringSecret.length() + stringSecret.length() % 2];
 
     final int offset = stringSecret.length() % 2;
```

Comment:
I think "offset" deserves a comment, if students try to understand the code they might not think of the complications of having a secret length odd or even